### ISSUE TITLE
Avoid boiler-plate when constructing a bokeh server programmatically

### ISSUE THREAD

**Issue by mrocklin** [bokeh/bokeh#6565]
In https://bokeh.github.io/blog/2017/6/29/simple_bokeh_server/ we discuss making a simple bokeh server from within Python.

```python
from bokeh.server.server import Server
from bokeh.application import Application
from bokeh.application.handlers.function import FunctionHandler
from bokeh.plotting import figure, ColumnDataSource

def make_document(doc):
    fig = figure(title='Line plot!', sizing_mode='scale_width')
    fig.line(x=[1, 2, 3], y=[1, 4, 9])

    doc.title = "Hello, world!"
    doc.add_root(fig)

apps = {'/': Application(FunctionHandler(make_document))}

server = Server(apps, port=5000)
server.start()
```

It might be worth thinking about how to reduce the amount of code, and in particular the number of foreign terms that a beginning user needs to understand to make this work.  In my ideal world the following might work:

```python
from bokeh.plotting import figure, ColumnDataSource, Server

def make_document(doc):
    fig = figure(title='Line plot!', sizing_mode='scale_width')
    fig.line(x=[1, 2, 3], y=[1, 4, 9])

    doc.title = "Hello, world!"
    doc.add_root(fig)

apps = {'/': make_document}

server = Server(apps, port=5000)
server.start()

**Comment by nleclercq** [comment]
@mrocklin I'm working on the same kind of application - i.e. "live-streaming, interactive visualizations that update with real-time data". My main goal is to provide our scientists with an experiments control environment mainly based on the jupyter notebook and bokeh. 

I'm notably evaluating several designs to embed bokeh into a jupyter notebook. What I have today is a [bokeh session model](https://github.com/nleclercq/jupyter-for-controls/blob/master/bokeh-data-streaming-for-notebook/session.py) allowing to easily spawn a bokeh server, to open some sessions and deal with periodic activity (i.e. periodic updates). The basic idea is simply to inherit from the `BokehSession` class and implement the `setup_document` and (optionally) the `periodic_callback` members. Everything else is handled behind the scene.  I personally like the idea of having my `app` implemented has a single python class. Bokeh could also propose such a high level API.

Here is a [BokehSession example](https://github.com/nleclercq/jupyter-for-controls/blob/master/03_about_columndatassource_patch.ipynb).

This post is just a way to let you know what I'm doing. I just wanted to share that with you cause our projects have the same kind of foundations. I'm also interested in any kind of review, remarks, advices from you or anyone else reading this. Thanks.

**Comment by bryevdv** [comment]
closed by #7077
